### PR TITLE
Support parallelized evaluation

### PR DESCRIPTION
**Pull request by changlin31**
Add support to parallelized evaluation (#6 ) with ordered distributed sampler in timm.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -202,7 +204,8 @@ def main(args):
     )
 
     data_loader_val = torch.utils.data.DataLoader(
-        dataset_val, batch_size=int(1.5 * args.batch_size),
+        dataset_val, sampler=sampler_val,
+        batch_size=int(1.5 * args.batch_size),
         shuffle=False, num_workers=args.num_workers,
```
Comment (fmassa) [inline main.py:209]: Can you also remove `shuffle=False`, so that it's clear that all shuffling is handled by the sampler?

File: main.py [hunk 2]
```diff
@@ -190,6 +191,7 @@ def main(args):
             sampler_train = torch.utils.data.DistributedSampler(
                 dataset_train, num_replicas=num_tasks, rank=global_rank, shuffle=True
             )
+        sampler_val = OrderedDistributedSampler(dataset_val, num_replicas=num_tasks, rank=global_rank)
```
Comment (fmassa) [inline main.py:194]: Even though now we always have `if True` in the branch, ideally you should also have the `sampler_val` be defined in the other branch (as a `SequentialSampler`)

**Review by fmassa — COMMENTED** [review]
Thanks for the PR!

I have a couple of minor comments in the code, and one larger comment that requires more thought, which I present as follows:

the implementation for `OrderedDistributedSampler` will repeat the elements of the val set so that it is a multiple of the `batch_size * num_replicas`. This means that the accuracy that we will be reporting will not be the same if we run on a single process or on multiple GPUs, which is something which we would ideally like to avoid.

I don't see any easy solutions for this without resorting to larger changes, so I would also like to get @TouvronHugo thoughts on this one.

File: main.py (hunk 2 above)
Comment (changlin31) [reply main.py:194]: Thank you for the review! Have changed them.

**Comment by changlin31** [comment]
Thanks for the review,

In my understanding, `OrderedDistributedSampler` only repeat elements so that it is evenly divisible by `num_replicas` but not `batch_size * num_replicas`. Thus, giving 50000 samples, the accuracy should remain the same when using 1, 2, 4, 8 or 16 GPUs, which covers most cases. Still, this is potentially a problem for other dataset or other `num_replicas`.

File: main.py [hunk 3]
```diff
@@ -190,8 +191,10 @@ def main(args):
...
```
Comment (fmassa) [inline main.py:194]:
```suggestion
        sampler_val = torch.utils.data.DistributedSampler(
            dataset_val, num_replicas=num_tasks, rank=global_rank, shuffle=False
        )
```

File: main.py [hunk 4]
```diff
@@ -15,6 +15,7 @@
 from pathlib import Path
 
 from timm.data import Mixup
+from timm.data.distributed_sampler import OrderedDistributedSampler
```
Comment (fmassa) [inline main.py:18]:
```suggestion
```

**Review by fmassa — APPROVED** [review]
Oh yes, you are right, `OrderedDistributedSampler` enforces divisibility by `num_replicas` only, so for `imagenet_val` using up to 16 GPUs should still be fine.

With this in mind, this seems like an ok trade-off to have for ImageNet, but we are going to be soon extending the code to support distillation and finetuning, so we will be using other datasets as well. @TouvronHugo what do you think?

One last nit, PyTorch's `DistributedSampler` now natively supports `shuffle=False` arguments, could you replace `OrderedDistributedSampler` with it? I've added suggested changes to the code to facilitate integrating those changes.

**Comment by fmassa** [comment]
Following discussions with @TouvronHugo , we decided that it would be preferable if we could instead have a command-line argument for parallel evaluation (default to False), and to emit a warning if the dataset size is not a multiple of the `num_replicas`, so that users are aware that they might have slightly different results.

What do you think?

**Review by fmassa — CHANGES_REQUESTED** [review]
Actually, after a second thought we also need to modify the `evaluate` function so that it synchronizes the metrics from different processes https://github.com/facebookresearch/deit/blob/dcd888df202050c14558e418ea7cbd9cd212cd3d/engine.py#L65
so this is a needed change before we are able to merge this PR.

**Comment by changlin31** [comment]
You are right, I missed the important synchronize. And I think adding a command-line argument is great. Will update shortly.

**Review by fmassa — APPROVED** [review]
Looks great, thanks a lot!
